src/messaging/ReliableMessageMgr.cpp

```diff
@@ -158,12 +158,6 @@ void ReliableMessageMgr::ExecuteActions()
             // Do not StartTimer, we will schedule the timer at the end of the timer handler.
             mRetransTable.ReleaseObject(entry);
 
-            // Dropping our entry marked the exchange as not having an un-acked
-            // message... but of course it _does_ have an un-acked message and
-            // we have just given up on waiting for the ack.
-
-            ec->GetReliableMessageContext()->SetMessageNotAcked(true);
```

Comment:
This kind-of raises a protocol question in the sense that, by not setting MessageNotAcked to true, we are stating that we are still open to receiving an acknowledgment of delivery via the response, although all our retransmits have been exhausted. This clearly is an indicator of misconfiguration of MRP timeouts or the application response timeout, isn't it? Ideally, if we have exhausted all retries, it should go up as an OnSendFailure() callback for the application to take a necessary action(of closing the exchange, etc) even if the ResponseTimeout is outstanding, isn't it? 
Ideally, the ResponseTimeout should be some function of the MRP intervals and number of retries with some additional buffer time, and not entirely decoupled from the MRP values.

MRP, at its level has a window within which it is expected to process acknowledgments to its sent messages and if that window passes, should it still process late response messages as Acks?

Reply:
> This kind-of raises a protocol question in the sense that, by not setting MessageNotAcked to true

There is no more "message not acked" flag, precisely because the name did not accurately reflect what the flag did.  What the flag did (and still does after this PR, under the new name) is track "We have not gotten an ack, and if we do get an ack for the relevant message counter value we will clear this flag".  That last "if ..." happens to be false once we reach this code, which is why setting the flag to true here, which was added recently, was wrong.

> This clearly is an indicator of misconfiguration of MRP timeouts or the application response timeout, isn't it?

Well...  The application response timeout can be quite long for a long-running operation.  MRP timeouts are quite short (4s for all resends) for devices that don't claim to be sleepy.  Realistically, the use of sleepy params or MRP timeouts is kind of broken, but that's a brokenness required by the spec, unfortunately, which has been reaised as a spec issue.

> Ideally, if we have exhausted all retries, it should go up as an OnSendFailure() callback for the application to take a necessary action(of closing the exchange, etc) even if the ResponseTimeout is outstanding, isn't it?

Honestly: that would break the world as things stand, as long as we are following the spec-required algorithm here.

> Ideally, the ResponseTimeout should be some function of the MRP intervals and number of retries with some additional buffer time, and not entirely decoupled from the MRP values.

It's not, generally, decoupled.  The default behavior is that the app provides how much time to allow on top of the MRP timeout.  So the response timeout is nearly always the MRP timeout + some app-defined constant.  

> should it still process late response messages as Acks

That is the big question, yes.  The current implementation does not, because it has no way to tell whether they are in a sane way.

Reply:
Ok. So, app response timeout is well correlated with MRP timeout which is good. 

However, OnSendFailure() is not implemented in our SDK. Curious what happens in scenarios where there is a critical issue in sending the message out(say the key is wrong and the message cannot be encrypted) and so all MRP retransmits fail. In that case, the application is forced to wait out the full ExchangeContext timeout when it could have been notified of a Send failure.

Also, regarding the ResponseTimeout, especially in Exchanges where there are multiple back-and-forth messaging, there is a distinction between the ResponseTimeout and a timeout for the entire Exchange. The former should, ideally capture the timeout for the current outstanding sent app message, whereas the latter would be a sum of all the individual back-and-forth. So, for each new send on the same exchange, the response timeout should be re-set for each new send. Not sure if we do this, or the Response timeout is essentially a timeout for all message send and receives.

>That is the big question, yes. The current implementation does not, because it has no way to tell whether they are in a sane way.
I think, we can ignore the MRPAcking part of it since that timeout has already passed, but still send the response to the application if it is still waiting and the Exchange is valid.

Reply:
> However, OnSendFailure() is not implemented in our SDK. 

Correct.

> Curious what happens in scenarios where there is a critical issue in sending the message out(say the key is wrong and the message cannot be encrypted) 

That's a synchronous failure from SendMessage.

> there is a distinction between the ResponseTimeout and a timeout for the entire Exchange

We have no concept of "timeout for the entire Exchange" right now.

>  for each new send on the same exchange, the response timeout should be re-set for each new send. 

Yep, that's what we do.  Or rather you can configure it once, and the timer is restarted for each send.  Or you can change it as you go (and CASE does that, in fact).